### PR TITLE
Add store and truncation from CrypTen

### PR DESCRIPTION
**Pull request by gmuraru**
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

## Affected Dependencies
List any dependencies that are required for this change.

## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

### CODE REVIEW

File: src/sympc/utils/mpc_utils.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    Returns:
+        The number of wraparounds
+    """
+
+    res = torch.zeros(size=share_list[0].size(), dtype=torch.long)
+    prev_share = share_list[0]
+    for cur_share in share_list[1:]:
+        next_share = cur_share + prev_share
+        res += ((prev_share < 0) & (cur_share > 0) & (next_share > 0)).long()
```
Comment (LaRiffle) [inline src/sympc/utils/mpc_utils.py:27]: You have a typo here!
https://github.com/facebookresearch/CrypTen/blob/81ae3df7f2025ec727e8012fb0703dd91da0e079/crypten/common/util.py#L36

**Review by LaRiffle — CHANGES_REQUESTED** [review]

File: src/sympc/protocol/spdz/spdz.py [hunk 2]
```diff
@@ -76,6 +150,9 @@ def mul_parties(
     :rtype: ShareTensor (in our case ShareTensorPointer)
     """
 
+    crypto_store = session.crypto_store
+    key_store = f"beaver_mul_{tuple(eps.shape)}_{tuple(delta.shape)}"
```
Comment (LaRiffle) [inline src/sympc/protocol/spdz/spdz.py:154]: This `key_store` sounds a bit hacky
I understand it is the equivalent of this: https://github.com/OpenMined/PySyft/blob/syft_0.2.x/syft/frameworks/torch/mpc/spdz.py#L84
So I wouldn't use strings I think!

File: src/sympc/store/crypto_store.py [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+    def populate_store(self, key_store: str, primitives: Union[List[Any], Any]) -> None:
+        if isinstance(primitives, list):
+            self.store[key_store].extend(primitives)
+        else:
+            self.store[key_store].append(primitives)
+
+    def get_primitives_from_store(
+        self, key_store: str, nr_instances: int = 1
+    ) -> List[Any]:
+        return self.store[key_store][: nr_instances + 1]
```
Comment (LaRiffle) [inline src/sympc/store/crypto_store.py:27]: i'd be very careful with `nr_instances`, from our experience it's better to not slice at all (because it's highly inefficient).
For triples for example, you could had a list of primitives that you pop out when needed (like here: https://github.com/OpenMined/PySyft/blob/syft_0.2.x/syft/frameworks/torch/mpc/primitives.py#L82) 
So smthg like
```
matmul_primitive = {shape_type: [triple_0, triple_1, ....]} # you pop out `triple_0`
```
For comparison primitives, where you have nr_instances > 1, better would even be to add the nr_instances int the key:
```
comp_primitive = {<shape>_<type>_<nr_instances>: [prim_0, prim_1, ....]} # you pop out `prim_0`
```

Also, here in your code, you don't delete the primitive once it's used?

File: src/sympc/tensor/mpc_tensor.py [hunk 4]
```diff
@@ -70,16 +79,24 @@ def __init__(
...
-                        break
+                        return
             else:
-                self.share_ptrs = []
+                nr_parties = len(self.session.parties)
+                tensor_type = self.session.tensor_type
+                shares = MPCTensor.generate_shares(secret, nr_parties, tensor_type)
+
+        if shares is not None and not ispointer(shares[0]):
+            shares = MPCTensor.distribute_shares(shares, self.session.parties)
```
Comment (LaRiffle) [inline src/sympc/tensor/mpc_tensor.py:89]: can you factorize this clause with L66 `if secret is not None:` ?

File: tests/sympc/tensor/test_mpc_tensor.py [hunk 5]
```diff
@@ -0,0 +1,214 @@
+import pytest
+import syft as sy
+import torch
+
+from sympc.session import Session
+from sympc.tensor import MPCTensor
+
+
+@pytest.fixture
```
Comment (LaRiffle) [inline tests/sympc/tensor/test_mpc_tensor.py:9]: duplicate with conftest?

File: src/sympc/__init__.py [hunk 6]
```diff
@@ -7,11 +7,20 @@
 
 from pkg_resources import DistributionNotFound, get_distribution
 
+try:
+    import syft
+except ImportError as e:
+    print("PySyft is needed to be able to use SyMPC")
+    raise (e)
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/__init__.py:14]: Will remove them - I do not need them! :D

File: src/sympc/protocol/beaver/beaver.py [hunk 7]
```diff
@@ -4,52 +4,76 @@
...
+) -> Tuple[Tuple[ShareTensor, ShareTensor, ShareTensor]]:
     """
     The Trusted Third Party (TTP) or Crypto Provider should provide this triples
     Currently, the one that orchestrates the communication provides those
     """
-    if op_str not in EXPECTED_OPS:
-        raise ValueError(f"{op_str} should be in {EXPECTED_OPS}")
+    a = ShareTensor(encoder_precision=0)
+    a.tensor = torch.zeros(a_shape, dtype=torch.long)
+    a_shares = MPCTensor.generate_shares(a, nr_parties, torch.long)
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/beaver/beaver.py:32]: it was meant for debugging purposes - will replace those with random ones

File: src/sympc/protocol/beaver/beaver.py [hunk 8]
```diff
@@ -4,52 +4,76 @@
...
+) -> Tuple[List[ShareTensor], List[ShareTensor], List[ShareTensor]]:
+    return _get_triples("mul", *args, **kwargs)
+
+
+@register_primitive_generator("beaver_wraps")
+def count_wraps_rand(
+    nr_parties: int, shape: Tuple[int]
+) -> Tuple[List[ShareTensor], List[ShareTensor]]:
+    r = ShareTensor(encoder_precision=0)
+    r.tensor = torch.zeros(shape, dtype=torch.long)
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/beaver/beaver.py:66]: Will replace this - still testing purposes

File: src/sympc/protocol/beaver/beaver.py [hunk 9]
```diff
@@ -4,52 +4,76 @@
...
+    r_shares = MPCTensor.generate_shares(r, nr_parties, torch.long)
+    wraps = count_wraps(r_shares)
+    for share in r_shares:
+        print(share)
+    print(wraps)
+
+    theta_r = ShareTensor(encoder_precision=0)
+    theta_r.tensor = wraps
+
+    theta_r_shares = MPCTensor.generate_shares(theta_r, nr_parties, torch.long)
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/beaver/beaver.py:77]: Hmm...maybe we can add a parameter, for the moment when we do the MPCTensor we also send the shares to the parties.
Using, ```MPCTensor.generate_shares``` would only split the secret, but the shares are kept local.

We can have:
```
x = ShareTensor(data = y, encoder_precision=0)
x_shares = MPCTensor.generate_shares(x, ...)
```

Or, I was thinking to have something like:
```x = MPCTensor.generate_shares(x)``` <--- and here to default the encoder_precision to 0

Do you have any ideas in mind? or are the ones above more appealing?

[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/beaver/beaver.py:77]: Currently, I replaced this with:
```
x = ShareTensor(data = y, encoder_precision=0)
x_shares = MPCTensor.generate_shares(x, ...)
```

I will push my changes this evening :D

File: src/sympc/protocol/spdz/spdz.py [hunk 10]
```diff
@@ -32,11 +36,25 @@ def mul_master(x: MPCTensor, y: MPCTensor, op_str: str) -> List[ShareTensor]:
...
+    shape_y = tuple(y.shape)
+
+    primitive_generator = CryptoPrimitiveProvider.get_generator("beaver_mul")
+    a_sh, b_sh, c_sh = primitive_generator(session.nr_parties, shape_x, shape_y)
+
+    key_store = f"beaver_{op_str}_{shape_x}_{shape_y}"
+    primitives = list(zip(a_sh, b_sh, c_sh))
+    CryptoPrimitiveProvider.transfer_primitives_to_parties(
+        key_store, primitives, session.session_ptrs
+    )
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/spdz/spdz.py:51]: Yeah...now that I am thinking of it - when you generate primitives, is not like you want them to be stored in the TTP, but the TTP should directly send them to the parties

File: src/sympc/protocol/spdz/spdz.py (hunk 2 above)
Comment (gmuraru) [reply src/sympc/protocol/spdz/spdz.py:154]: Hmm - yep, but I was thinking that we need a way to "register" custom "setters" and "getters" for primitives? - I can try something different - will update the PR and ping you when is ready :D

File: src/sympc/store/crypto_store.py (hunk 3 above)
Comment (gmuraru) [reply src/sympc/store/crypto_store.py:27]: Yep - I should also delete it after using it.

Do you have any idea if using ```numpy.views``` would provide a better runtime?

File: src/sympc/protocol/beaver/beaver.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (LaRiffle) [reply src/sympc/protocol/beaver/beaver.py]: Ok cool, it's better!
But there is still this thing where I think "Ok I have generated a share (because I used "ShareTensor"), and then I use this in generate_shares ?!" Maybe something more direct like you suggested would be more readable for the new readers like me :D 
```python
x_shares = MPCTensor.generate_shares(x, encoder_precision=0, ...)
```

File: src/sympc/store/crypto_store.py (hunk 3 above)
Comment (LaRiffle) [reply src/sympc/store/crypto_store.py]: not sure :/

File: tests/sympc/tensor/test_mpc_tensor.py (hunk 5 above)
Comment (gmuraru) [reply tests/sympc/tensor/test_mpc_tensor.py:9]: Yep - I will add something on conftest that generates a specific number of workers

File: src/sympc/tensor/mpc_tensor.py (hunk 4 above)
Comment (gmuraru) [reply src/sympc/tensor/mpc_tensor.py:89]: Hmm...looked again into this, this is for the case we don't get a ```secret``` and get a list of shares (that are local) and we need to distribute them.

**Review by LaRiffle — CHANGES_REQUESTED** [review]
Awesome work!
I have left small questions and remarks :)

File: src/sympc/protocol/beaver/beaver.py [hunk 11]
```diff
@@ -4,55 +4,176 @@
...
+    rand_val = torch.empty(size=shape, dtype=torch.long).random_(
+        generator=ttp_generator
+    )
+    r = ShareTensor(data=rand_val, encoder_precision=0)
+
+    r_shares = MPCTensor.generate_shares(r, nr_parties, torch.long)
+    wraps = count_wraps(r_shares)
+
+    theta_r = ShareTensor(data=wraps, encoder_precision=0)
+    theta_r.tensor = wraps
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/beaver/beaver.py:175]: yes

File: src/sympc/protocol/spdz/spdz.py [hunk 12]
```diff
@@ -48,21 +68,98 @@ def mul_master(x: MPCTensor, y: MPCTensor, op_str: str) -> List[ShareTensor]:
...
+    From CrypTen
+    Privately computes the number of wraparounds for a set a shares
+
+    To do so, we note that:
+        [theta_x] = theta_z + [beta_xr] - [theta_r] - [eta_xr]
+
+    Where:
+        [theta_i] is the wraps for a variable i
+        [beta_ij] is the differential wraps for variables i and j
+        [eta_ij]  is the plaintext wraps for variables i and j
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/spdz/spdz.py:142]: Yep - I copy-pasted this from the CrypTen repository.

File: src/sympc/store/crypto_store.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
+# stdlib
+from typing import Any
+from typing import Callable
+from typing import Dict
+from typing import Iterable
+from typing import List
+
+
+class CryptoStore:
+    _FUNC_ADD_STORE: Dict[Any, Callable] = {}
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/store/crypto_store.py:10]: Hmm...I have seen it used for global vars, I can change it to lowercase :D

File: src/sympc/tensor/mpc_tensor.py [hunk 14]
```diff
@@ -148,40 +169,37 @@ def generate_przs(
...
 
         for _ in range(nr_parties - 1):
-            rand_value = torch.empty(size=shape, dtype=torch.long).random_(
+            rand_value = torch.empty(size=shape, dtype=tensor_type).random_(
                 generator=generator
             )
-            share = ShareTensor(session=session)
-
-            # Add the share after such that we do not encode it
+            share = ShareTensor(session=secret.session)
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/tensor/mpc_tensor.py:196]: In this case no, because we do not want it to be encoded

File: src/sympc/utils/mpc_utils.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
+# stdlib
+from typing import List
+
+# third party
+# TODO: Remove this
+import torch
+
+
+def count_wraps(share_list: List[torch.tensor]) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/utils/mpc_utils.py:9]: I replaced the other one to send a list of ```torch.Tensors``` - I was thinking to keep this ```mpc_utils.py``` file cleaner such that we do not send something specific from our implementation + at one point we also send ```torch.tensors```

File: src/sympc/protocol/beaver/beaver.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/protocol/beaver/beaver.py]: I opened an issue for [it](https://github.com/OpenMined/SyMPC/issues/46) :( - I think this PR is getting very very massive.

File: src/sympc/tensor/mpc_tensor.py [hunk 16]
```diff
@@ -348,6 +389,23 @@ def __apply_public_op(
         result = MPCTensor(shares=shares, session=self.session)
         return result
 
+    @staticmethod
+    def __get_shape(
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/tensor/mpc_tensor.py:395]: yes

File: src/sympc/tensor/mpc_tensor.py [hunk 17]
```diff
@@ -61,32 +60,54 @@ def __init__(
...
 
-        elif shares is not None:
-            self.share_ptrs = shares
+        if not ispointer(shares[0]):
+            shares = MPCTensor.distribute_shares(shares, self.session.parties)
+
+        if shape is not None:
+            self.shape = shape
+
+        self.share_ptrs = shares
```
[the remark this answers is not in the document]
Comment (gmuraru) [reply src/sympc/tensor/mpc_tensor.py:103]: ```self.shares_ptrs = shares``` can also reach this point if we pass in ```shares``` and not the ```secret```

File: src/sympc/store/crypto_store.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (LaRiffle) [reply src/sympc/store/crypto_store.py]: no that's fine! just a question :)

**Review by LaRiffle — APPROVED** [review]
All good for me! 🙌
